base/src/main/java/org/arend/typechecking/visitor/DesugarVisitor.java

```diff
@@ -233,6 +240,47 @@ public Void visitClass(Concrete.ClassDefinition def, Void params) {
     return visitApp((Concrete.ReferenceExpression) expr.getFunction(), expr.getArguments(), expr, true);
   }
 
+  private static void convertAppHoles(Concrete.AppExpression expr, List<Concrete.Parameter> parameters) {
+    Concrete.Expression originalFunc = expr.getFunction();
+    if (originalFunc instanceof Concrete.ApplyHoleExpression) {
+      Object data = originalFunc.getData();
+      LocalReferable ref = new LocalReferable("p" + parameters.size());
+      parameters.add(new Concrete.NameParameter(data, true, ref));
+      expr.setFunction(new Concrete.ReferenceExpression(data, ref));
+    }
+    boolean isOp = false;
+    if (originalFunc instanceof Concrete.ReferenceExpression) {
+      Referable referable = originalFunc.getUnderlyingReferable();
+      if (referable instanceof GlobalReferable) {
+        Precedence precedence = ((GlobalReferable) referable).getPrecedence();
+        isOp = precedence.isInfix;
+      }
+    }
+    for (Concrete.Argument argument : expr.getArguments())
+      if (argument.expression instanceof Concrete.ApplyHoleExpression) {
+        Object data = argument.expression.getData();
+        LocalReferable ref = new LocalReferable("p" + parameters.size());
+        parameters.add(new Concrete.NameParameter(data, argument.isExplicit(), ref));
+        argument.expression = new Concrete.ReferenceExpression(data, ref);
+      } else if (isOp && argument.expression instanceof Concrete.AppExpression) {
+        convertAppHoles((Concrete.AppExpression) argument.expression, parameters);
+      } else if (isOp && argument.expression instanceof Concrete.ProjExpression) {
+        Concrete.ProjExpression proj = (Concrete.ProjExpression) argument.expression;
+        if (!(proj.expression instanceof Concrete.ApplyHoleExpression)) return;
+        Object data = proj.expression.getData();
+        LocalReferable ref = new LocalReferable("p" + parameters.size());
+        parameters.add(new Concrete.NameParameter(data, true, ref));
+        proj.expression = new Concrete.ReferenceExpression(data, ref);
+      }
+  }
+
+  @Override
+  public Concrete.Expression visitApplyHole(Concrete.ApplyHoleExpression expr, Void params) {
+    // TODO: better error type
+    myErrorReporter.report(new GeneralError(GeneralError.Level.ERROR, "`__` not allowed here"));
```

Comment:
@valis Is there a better way?

Reply:
It is better to use `TypecheckingError` at this stage.

Reply:
Fixed